File: cubix.py
import sys


class Cubix:
    def __init__(self, size=3):
        self.n = size  # typically 3
        self.mid = self.n//2
        self.faces = {
            'F': [[0] * self.n for _ in range(self.n)],
            'B': [[0] * self.n for _ in range(self.n)],
            'U': [[0] * self.n for _ in range(self.n)],
            'D': [[0] * self.n for _ in range(self.n)],
            'L': [[0] * self.n for _ in range(self.n)],
            'R': [[0] * self.n for _ in range(self.n)],
        }
        self.position_map = {
            'F': 'F',
            'B': 'B',
            'U': 'U',
            'D': 'D',
            'L': 'L',
            'R': 'R',
        }
        self.x, self.y = self.n//2, self.n//2

    def rotate_cube_faces(self, axis, clockwise):
        p = self.position_map
        if axis == 'X':
            if clockwise:
                p['F'], p['D'], p['B'], p['U'] = p['D'], p['B'], p['U'], p['F']
            else:
                p['F'], p['U'], p['B'], p['D'] = p['U'], p['B'], p['D'], p['F']
        elif axis == 'Y':
            if clockwise:
                p['L'], p['B'], p['R'], p['F'] = p['B'], p['R'], p['F'], p['L']
            else:
                p['L'], p['F'], p['R'], p['B'] = p['F'], p['R'], p['B'], p['L']
        elif axis == 'Z':
            if clockwise:
                p['U'], p['R'], p['D'], p['L'] = p['L'], p['U'], p['R'], p['D']
            else:
                p['U'], p['L'], p['D'], p['R'] = p['R'], p['U'], p['L'], p['D']
        else:
            print(f"Unknown axis: {axis}")
            sys.exit(1)

    def rotate_face(self, face, clockwise):
        F, R, B, L, U, D = self.faces["F"], self.faces["R"], self.faces["B"], self.faces["L"], self.faces["U"], self.faces["D"]
        n = self.n

        if clockwise:
            self.faces[face] = [list(row) for row in zip(*self.faces[face][::-1])]
            if face == "U":
                tmp = F[0][:]
                F[0] = L[0][:]
                L[0] = B[0][:]
                B[0] = R[0][:]
                R[0] = tmp
            elif face == "D":
                tmp = F[n-1][:]
                F[n-1] = R[n-1][:]
                R[n-1] = B[n-1][:]
                B[n-1] = L[n-1][:]
                L[n-1] = tmp
            elif face == "L":
                tmp = [U[i][0] for i in range(n)]
                for i in range(n):
                    U[i][0] = B[n-1-i][n-1]
                    B[n-1-i][n-1] = D[i][0]
                    D[i][0] = F[i][0]
                    F[i][0] = tmp[i]
            elif face == "R":
                tmp = [U[i][n-1] for i in range(n)]
                for i in range(n):
                    U[i][n-1] = F[i][n-1]
                    F[i][n-1] = D[i][n-1]
                    D[i][n-1] = B[n-1-i][0]
                    B[n-1-i][0] = tmp[i]
            elif face == "F":
                tmp = U[n-1][:]
                for i in range(n):
                    U[n-1][i] = L[n-1-i][n-1]
                    L[n-1-i][n-1] = D[0][n-1-i]
                    D[0][n-1-i] = R[i][0]
                    R[i][0] = tmp[i]
            elif face == "B":
                tmp = U[0][:]
                for i in range(n):
                    U[0][i] = R[i][n-1]
                    R[i][n-1] = D[n-1][n-1-i]
                    D[n-1][n-1-i] = L[n-1-i][0]
                    L[n-1-i][0] = tmp[i]

        else:
            self.faces[face] = [list(row) for row in zip(*self.faces[face])][::-1]
            if face == "U":
                tmp = F[0][:]
                F[0] = R[0][:]
                R[0] = B[0][:]
                B[0] = L[0][:]
                L[0] = tmp
            elif face == "D":
                tmp = F[n-1][:]
                F[n-1] = L[n-1][:]
                L[n-1] = B[n-1][:]
                B[n-1] = R[n-1][:]
                R[n-1] = tmp
            elif face == "L":
                tmp = [U[i][0] for i in range(n)]
                for i in range(n):
                    U[i][0] = F[i][0]
                    F[i][0] = D[i][0]
                    D[i][0] = B[n-1-i][n-1]
                    B[n-1-i][n-1] = tmp[i]
            elif face == "R":
                tmp = [U[i][n-1] for i in range(n)]
                for i in range(n):
                    U[i][n-1] = B[n-1-i][0]
                    B[n-1-i][0] = D[i][n-1]
                    D[i][n-1] = F[i][n-1]
                    F[i][n-1] = tmp[i]
            elif face == "F":
                tmp = U[n-1][:]
                for i in range(n):
                    U[n-1][i] = R[i][0]
                    R[i][0] = D[0][n-1-i]
                    D[0][n-1-i] = L[n-1-i][n-1]
                    L[n-1-i][n-1] = tmp[i]
            elif face == "B":
                tmp = U[0][:]
                for i in range(n):
                    U[0][i] = L[n-1-i][0]
                    L[n-1-i][0] = D[n-1][n-1-i]
                    D[n-1][n-1-i] = R[i][n-1]
                    R[i][n-1] = tmp[i]

    def increment_cell(self):
        front = self.position_map['F']
        self.faces[front][self.x][self.y] += 1

    def decrement_cell(self):
        front = self.position_map['F']
        self.faces[front][self.x][self.y] -= 1

    def output_cell(self):
        front = self.position_map['F']
        print(self.faces[front][self.x][self.y])

    def input_cell(self):
        front = self.position_map['F']
        self.faces[front][self.x][self.y] = int(input())

    def dump_front_face(self):
        front = self.faces(self.position_map['F'])
        print(self.faces[front])

    def dump_cube_state(self):
        F, R, B, L, U, D = self.faces["F"], self.faces["R"], self.faces["B"], self.faces["L"], self.faces["U"], self.faces["D"]
        def row_str(row): return " ".join(str(x) for x in row)

        def padded(rows, letter):
            return ["            [" + letter + "] " + row_str(row) for row in rows]

        print("\n".join([
            *padded(U, "U"),
            "",
            "[L] " + row_str(L[0]) + "   [F] " + row_str(F[0]) + "   [R] " + row_str(R[0]) + "   [B] " + row_str(B[0]),
            "[L] " + row_str(L[1]) + "   [F] " + row_str(F[1]) + "   [R] " + row_str(R[1]) + "   [B] " + row_str(B[1]),
            "[L] " + row_str(L[2]) + "   [F] " + row_str(F[2]) + "   [R] " + row_str(R[2]) + "   [B] " + row_str(B[2]),
            "",
            *padded(D, "D")
        ]))

    def move_left(self):
        if self.x >= 0:
            self.x -= 1

    def move_right(self):
        if self.x < self.n:
            self.x += 1

    def move_up(self):
        if self.y >= 0:
            self.y -= 1

    def move_down(self):
        if self.y < self.n:
            self.y += 1


def tokenize(code_str):
    loops = []
    tokens = []
    i = 0
    while i < len(code_str):
        if code_str[i] == " " or code_str[i] == ")" or code_str[i] == "\n":
            i += 1
            continue
        elif code_str[i] == "[":
            i += 1  # skip '['
            loop_number = ""
            while i < len(code_str) and code_str[i] != "]":
                loop_number += code_str[i]
                i += 1
            i += 1  # skip ']'
            loops.append({'type': 'loop', 'count': int(loop_number)})
            continue
        elif code_str[i] == "(":
            i += 1  # skip '('
            loop_code = ""
            while i < len(code_str) and code_str[i] != ")":
                loop_code += code_str[i]
                i += 1
            i += 1  # skip ')'
            loops[-1]['body'] = tokenize(loop_code)
            tokens.append(loops[-1])
            continue
        elif i + 1 < len(code_str) and code_str[i+1] == "'":
            tokens.append(code_str[i] + "'")
            i += 2
        else:
            tokens.append(code_str[i])
            i += 1
    return tokens


def parse_and_run(code, cubix):
    code = tokenize(code)  # valid commands
    pc = 0  # program counter
    code_len = len(code)

    while pc < code_len:
        cmd = code[pc]
        if cmd in ["X'", "Y'", "Z'"]:
            cubix.rotate_cube_faces(cmd[0], clockwise=False)
        elif cmd in ["X", "Y", "Z"]:
            cubix.rotate_cube_faces(cmd, clockwise=True)
        elif cmd in ["F'", "B'", "U'", "D'", "L'", "R'"]:
            cubix.rotate_face(cmd[0], clockwise=False)
        elif cmd in ["F", "B", "U", "D", "L", "R"]:
            cubix.rotate_face(cmd, clockwise=True)
        elif cmd == "+":
            cubix.increment_cell()
        elif cmd == "-":
            cubix.decrement_cell()
        elif cmd == ".":
            cubix.output_cell()
        elif cmd == ",":
            cubix.input_cell()
        elif cmd == "p":
            cubix.dump_front_face()
        elif cmd == "d":
            cubix.dump_cube_state()
        elif cmd == '<':
            cubix.move_left()
        elif cmd == '>':
            cubix.move_right()
        elif cmd == '^':
            cubix.move_up()
        elif cmd == 'v':
            cubix.move_down()
        elif cmd == 'p':
            print(f"Cursor: ({cubix.x}, {cubix.y})")
        elif isinstance(cmd, dict) and cmd.get('type') == 'loop':
            for _ in range(cmd['count']):
                parse_and_run(cmd['body'], cubix)
        else:
            print(f"Unknown command: {cmd}")
            sys.exit(1)
        pc += 1


def main():
    if len(sys.argv) != 2:
        print("Usage: python3 cubix.py program.cbx")
        sys.exit(1)

    with open(sys.argv[1]) as f:
        code = f.read()
    cubix = Cubix(size=3)
    parse_and_run(code, cubix)
    print()  # newline at the end


if __name__ == "__main__":
    main()

File: cubix.js
class Cubix {
    constructor(size = 3) {
      this.n = size;
      this.mid = Math.floor(this.n / 2);
      this.faces = {
        F: this.makeFace(),
        B: this.makeFace(),
        U: this.makeFace(),
        D: this.makeFace(),
        L: this.makeFace(),
        R: this.makeFace(),
      };
      this.positionMap = {
        F: "F",
        B: "B",
        U: "U",
        D: "D",
        L: "L",
        R: "R",
      };
      this.x = this.mid;
      this.y = this.mid;
    }
  
    makeFace() {
      return Array(this.n)
        .fill(null)
        .map(() => Array(this.n).fill(0));
    }
  
    rotateCubeFaces(axis, clockwise) {
      const p = this.positionMap;
      if (axis === "X") {
        if (clockwise) {
          [p.F, p.D, p.B, p.U] = [p.D, p.B, p.U, p.F];
        } else {
          [p.F, p.U, p.B, p.D] = [p.U, p.B, p.D, p.F];
        }
      } else if (axis === "Y") {
        if (clockwise) {
          [p.L, p.B, p.R, p.F] = [p.B, p.R, p.F, p.L];
        } else {
          [p.L, p.F, p.R, p.B] = [p.F, p.R, p.B, p.L];
        }
      } else if (axis === "Z") {
        if (clockwise) {
          [p.U, p.R, p.D, p.L] = [p.L, p.U, p.R, p.D];
        } else {
          [p.U, p.L, p.D, p.R] = [p.R, p.U, p.L, p.D];
        }
      } else {
        throw new Error(`Unknown axis: ${axis}`);
      }
    }
  
    rotateFace(face, clockwise) {
      const F = this.faces.F,
        R = this.faces.R,
        B = this.faces.B,
        L = this.faces.L,
        U = this.faces.U,
        D = this.faces.D;
      const n = this.n;
  
      // Rotate the face matrix itself
      if (clockwise) {
        this.faces[face] = this.faces[face]
          .map((_, i) => this.faces[face].map((row) => row[i]))
          .map((row) => row.reverse());
      } else {
        this.faces[face] = this.faces[face]
          .map((_, i) => this.faces[face].map((row) => row[i]))
          .reverse();
      }
  
      // Rotate adjacent edges depending on face and direction
      if (clockwise) {
        if (face === "U") {
          let tmp = F[0].slice();
          F[0] = L[0].slice();
          L[0] = B[0].slice();
          B[0] = R[0].slice();
          R[0] = tmp;
        } else if (face === "D") {
          let tmp = F[n - 1].slice();
          F[n - 1] = R[n - 1].slice();
          R[n - 1] = B[n - 1].slice();
          B[n - 1] = L[n - 1].slice();
          L[n - 1] = tmp;
        } else if (face === "L") {
          let tmp = U.map((row) => row[0]);
          for (let i = 0; i < n; i++) {
            U[i][0] = B[n - 1 - i][n - 1];
            B[n - 1 - i][n - 1] = D[i][0];
            D[i][0] = F[i][0];
            F[i][0] = tmp[i];
          }
        } else if (face === "R") {
          let tmp = U.map((row) => row[n - 1]);
          for (let i = 0; i < n; i++) {
            U[i][n - 1] = F[i][n - 1];
            F[i][n - 1] = D[i][n - 1];
            D[i][n - 1] = B[n - 1 - i][0];
            B[n - 1 - i][0] = tmp[i];
          }
        } else if (face === "F") {
          let tmp = U[n - 1].slice();
          for (let i = 0; i < n; i++) {
            U[n - 1][i] = L[n - 1 - i][n - 1];
            L[n - 1 - i][n - 1] = D[0][n - 1 - i];
            D[0][n - 1 - i] = R[i][0];
            R[i][0] = tmp[i];
          }
        } else if (face === "B") {
          let tmp = U[0].slice();
          for (let i = 0; i < n; i++) {
            U[0][i] = R[i][n - 1];
            R[i][n - 1] = D[n - 1][n - 1 - i];
            D[n - 1][n - 1 - i] = L[n - 1 - i][0];
            L[n - 1 - i][0] = tmp[i];
          }
        }
      } else {
        if (face === "U") {
          let tmp = F[0].slice();
          F[0] = R[0].slice();
          R[0] = B[0].slice();
          B[0] = L[0].slice();
          L[0] = tmp;
        } else if (face === "D") {
          let tmp = F[n - 1].slice();
          F[n - 1] = L[n - 1].slice();
          L[n - 1] = B[n - 1].slice();
          B[n - 1] = R[n - 1].slice();
          R[n - 1] = tmp;
        } else if (face === "L") {
          let tmp = U.map((row) => row[0]);
          for (let i = 0; i < n; i++) {
            U[i][0] = F[i][0];
            F[i][0] = D[i][0];
            D[i][0] = B[n - 1 - i][n - 1];
            B[n - 1 - i][n - 1] = tmp[i];
          }
        } else if (face === "R") {
          let tmp = U.map((row) => row[n - 1]);
          for (let i = 0; i < n; i++) {
            U[i][n - 1] = B[n - 1 - i][0];
            B[n - 1 - i][0] = D[i][n - 1];
            D[i][n - 1] = F[i][n - 1];
            F[i][n - 1] = tmp[i];
          }
        } else if (face === "F") {
          let tmp = U[n - 1].slice();
          for (let i = 0; i < n; i++) {
            U[n - 1][i] = R[i][0];
            R[i][0] = D[0][n - 1 - i];
            D[0][n - 1 - i] = L[n - 1 - i][n - 1];
            L[n - 1 - i][n - 1] = tmp[i];
          }
        } else if (face === "B") {
          let tmp = U[0].slice();
          for (let i = 0; i < n; i++) {
            U[0][i] = L[n - 1 - i][0];
            L[n - 1 - i][0] = D[n - 1][n - 1 - i];
            D[n - 1][n - 1 - i] = R[i][n - 1];
            R[i][n - 1] = tmp[i];
          }
        }
      }
    }
  
    incrementCell() {
      const front = this.positionMap.F;
      this.faces[front][this.x][this.y]++;
    }
  
    decrementCell() {
      const front = this.positionMap.F;
      this.faces[front][this.x][this.y]--;
    }
  
    outputCell() {
      const front = this.positionMap.F;
      // For playground, let's output as a character
      return String.fromCharCode(this.faces[front][this.x][this.y]);
    }
  
    inputCell(charCode) {
      const front = this.positionMap.F;
      this.faces[front][this.x][this.y] = charCode;
    }
  
    moveLeft() {
      if (this.y > 0) this.y--;
    }
  
    moveRight() {
      if (this.y < this.n - 1) this.y++;
    }
  
    moveUp() {
      if (this.x > 0) this.x--;
    }
  
    moveDown() {
      if (this.x < this.n - 1) this.x++;
    }
  
    dumpFrontFace() {
      const front = this.faces[this.positionMap.F];
      return front.map((row) => row.join(" ")).join("\n");
    }
  
    dumpCubeState() {
      const { F, R, B, L, U, D } = this.faces;
      const rowStr = (row) => row.join(" ");
      const pad = (rows, letter) =>
        rows.map((row) => `            [${letter}] ${rowStr(row)}`);
  
      return [
        ...pad(U, "U"),
        "",
        `[L] ${rowStr(L[0])}   [F] ${rowStr(F[0])}   [R] ${rowStr(R[0])}   [B] ${rowStr(B[0])}`,
        `[L] ${rowStr(L[1])}   [F] ${rowStr(F[1])}   [R] ${rowStr(R[1])}   [B] ${rowStr(B[1])}`,
        `[L] ${rowStr(L[2])}   [F] ${rowStr(F[2])}   [R] ${rowStr(R[2])}   [B] ${rowStr(B[2])}`,
        "",
        ...pad(D, "D"),
      ].join("\n");
    }
  }
  
  // Tokenizer — same logic as Python version
  function tokenize(codeStr) {
    const loops = [];
    const tokens = [];
    let i = 0;
    while (i < codeStr.length) {
      if (codeStr[i] === " " || codeStr[i] === ")" || codeStr[i] === "\n") {
        i++;
        continue;
      } else if (codeStr[i] === "[") {
        i++;
        let loopNumber = "";
        while (i < codeStr.length && codeStr[i] !== "]") {
          loopNumber += codeStr[i];
          i++;
        }
        i++; // skip ']'
        loops.push({ type: "loop", count: parseInt(loopNumber, 10) });
        continue;
      } else if (codeStr[i] === "(") {
        i++;
        let loopCode = "";
        while (i < codeStr.length && codeStr[i] !== ")") {
          loopCode += codeStr[i];
          i++;
        }
        i++; // skip ')'
        loops[loops.length - 1].body = tokenize(loopCode);
        tokens.push(loops[loops.length - 1]);
        continue;
      } else if (i + 1 < codeStr.length && codeStr[i + 1] === "'") {
        tokens.push(codeStr[i] + "'");
        i += 2;
      } else {
        tokens.push(codeStr[i]);
        i++;
      }
    }
    return tokens;
  }
  
  // Main interpreter
  function parseAndRun(code, cubix, inputFn, outputFn) {
    const tokens = tokenize(code);
    let pc = 0;
  
    while (pc < tokens.length) {
      const cmd = tokens[pc];
  
      if (cmd === "X'" || cmd === "Y'" || cmd === "Z'") {
        cubix.rotateCubeFaces(cmd[0], false);
      } else if (cmd === "X" || cmd === "Y" || cmd === "Z") {
        cubix.rotateCubeFaces(cmd, true);
      } else if (["F'", "B'", "U'", "D'", "L'", "R'"].includes(cmd)) {
        cubix.rotateFace(cmd[0], false);
      } else if (["F", "B", "U", "D", "L", "R"].includes(cmd)) {
        cubix.rotateFace(cmd, true);
      } else if (cmd === "+") {
        cubix.incrementCell();
      } else if (cmd === "-") {
        cubix.decrementCell();
      } else if (cmd === ".") {
        const val = cubix.outputCell();
        outputFn(String.fromCharCode(val));
      } else if (cmd === ",") {
        const inputChar = inputFn();
        cubix.inputCell(inputChar.charCodeAt(0));
      } else if (cmd === "p") {
        cubix.dumpFrontFace();
      } else if (cmd === "d") {
        cubix.dumpCubeState();
      } else if (cmd === "<") {
        cubix.moveLeft();
      } else if (cmd === ">") {
        cubix.moveRight();
      } else if (cmd === "^") {
        cubix.moveUp();
      } else if (cmd === "v") {
        cubix.moveDown();
      } else if (typeof cmd === "object" && cmd.type === "loop") {
        for (let i = 0; i < cmd.count; i++) {
          parseAndRun(cmd.body, cubix, inputFn, outputFn);
        }
      } else {
        throw new Error(`Unknown command: ${cmd}`);
      }
  
      pc++;
    }
  }